backend/src/hatchling/builders/binary.py

```diff
@@ -111,74 +164,69 @@ def build_bootstrap(
         import shutil
         import tempfile
 
-        cargo_path = os.environ.get('CARGO', '')
-        if not cargo_path:
-            if not shutil.which('cargo'):
-                message = 'Executable `cargo` could not be found on PATH'
-                raise OSError(message)
-
-            cargo_path = 'cargo'
-
         app_dir = os.path.join(directory, self.PLUGIN_NAME)
         if not os.path.isdir(app_dir):
             os.makedirs(app_dir)
 
-        on_windows = sys.platform == 'win32'
-        base_env = dict(os.environ)
-        base_env['PYAPP_PROJECT_NAME'] = self.metadata.name
-        base_env['PYAPP_PROJECT_VERSION'] = self.metadata.version
-
-        if self.config.python_version:
-            base_env['PYAPP_PYTHON_VERSION'] = self.config.python_version
+        for output_spec in self.config.outputs:
+            env_vars = {  # merge options from the parent options table and the build target options table
+                **self.config.env_vars,
+                **output_spec.get('env-vars', {}),
+            }
+            # Format values with context
+            context = self.metadata.context
+            env_vars = {k: context.format(v) for k, v in env_vars.items()}
+
+            with EnvVars(env_vars):
+                cargo_path = os.environ.get('CARGO', '')
+                if not cargo_path:
+                    if not shutil.which('cargo'):
+                        message = 'Executable `cargo` could not be found on PATH'
+                        raise OSError(message)
+
+                    cargo_path = 'cargo'
+
+                on_windows = sys.platform == 'win32'
+                base_env = dict(os.environ)
+                base_env['PYAPP_PROJECT_NAME'] = self.metadata.name
```

Comment:
I have tested the proposed changes and this **does not** fix the issue defined in #1436. The project is not actually installed within the PyApp binary, resulting in a 

According to the [PyApp Docs](https://ofek.dev/pyapp/latest/config/project/#sources) there are 3 ways to define the project path. Here, `PYAPP_PROJECT_NAME` and `PYAPP_PROJECT_VERSION` are defined, meaning you are using the [Identifier](https://ofek.dev/pyapp/latest/config/project/#identifier) configuration which means **"the package will be installed from a package index like PyPI"**. This is not what we want, as the package is local and not necessarily on PyPi.

Rather than defining these two variables here, the builder should utilize the [Embedding](https://ofek.dev/pyapp/latest/config/project/#embedding) config method:
- Build an `sdist` or `wheel` in a temporary directory
- Set `PYAPP_PROJECT_PATH` to point to this file
  - Name and version are derived from the 

See [my comment on the related issue](https://github.com/pypa/hatch/issues/1436#issuecomment-2311391337) for how I got around this issue by building and pointing to an sdist.